dapr/proto/runtime/v1/dapr.proto

```diff
@@ -148,6 +153,45 @@ message GetStateRequest {
   map<string, string> metadata = 4;
 }
 
+// ComponentHealthResponseItem represents health of a single component item in the response
+message ComponentHealthResponseItem {
+  enum Status {
+    OK = 0;
+    NOT_OK = 1;
+    UNDEFINED = 2;
+  }
+
+  // The name of component
+  optional string component_name = 1;
```

Comment:
The first 3 should be required, not? I can see how error_code and message are optional but why would there be an entry without component_name, type or status?

Reply:
In case of enquiry for a single component, it is returning only `status` and not component_name and type.
Now, when I made everything else than `status` optional, so status was getting printed as first thing even in case of enquiring for ALL components. So, I made status as optional too fix ordering.

Reply:
This is odd. We should have the same output regardless the response being for one or many components.

Reply:
IMHO I think we should have two methods, one for all components (RESTful-ly, `GET /healthz/component`) and one for a single component (in REST, `GET /healthz/component/componentName`).

The reasons would be:

1. This is more RESTful than using querystring args
1. When requesting a single component, the response should be just the object. When requesting all components, the response should be an array.
1. When requesting a single component, this way allows us to return a status code indicating if the component is good (200) or if there's an error.

Reply:
> This is odd. We should have the same output regardless the response being for one or many components.

I have been intentionally removing `component_name` and `type` intentionally when only one component is enquired for. As it is the component which was already sent in request, so didn't want to show it again in response. I can change that to start showing component_name and type too, if that is what we want. I have put all the responses as per current in https://github.com/dapr/dapr/pull/4758#issuecomment-1375774333

Reply:
> IMHO I think we should have two methods, one for all components (RESTful-ly, `GET /healthz/component`) and one for a single component (in REST, `GET /healthz/component/componentName`).
> 
> The reasons would be:
> 
> 1. This is more RESTful than using querystring args
> 2. When requesting a single component, the response should be just the object. When requesting all components, the response should be an array.
> 3. When requesting a single component, this way allows us to return a status code indicating if the component is good (200) or if there's an error.

I agree with @ItalyPaleAle, we should try and follow the REST API guidelines here. 

One small suggestion 
`GET /healthz/components` and `GET /healthz/components/{componentName}`. Let's have the resource name as plural. 

Also for these two APIs we can have two responses, one which only has a single item (the individual component one), and one the array of items. 

For gRPC, should we use `ListComponentHealthAlpha1` for list heath status of all components and `GetComponentHealthAlpha1(componentName)` to get the health of a particular component? 

cc @ItalyPaleAle 

@dapr/maintainers-dapr thoughts?